### ISSUE TITLE
Suggestions for Offline

### ISSUE THREAD

**Issue by nitrag** [rmnblm/Mapbox-iOS-Examples#3]
Just some thoughts, I'll eventually be needing these and will contribute (unless you can beat me to it 👍 )
- [x] Start/Stop/Pause/Resume
- [x] Cancel
- [x] Show percentage label
- [ ] Display error if not internet connection (can't download if offline)
- [ ] Display error if download already in progress (don't allow downloading twice)
- [ ] Enable in background thread? So download will complete when app is minimized
- [ ] Display estimation of how much data will be downloaded (megabytes)
- [ ] Max zoom? For raster satellite layer (if that's even possible) it might be too insane to download all the tiles from visible -> infinity (is that how it works now?)

**Comment by rmnblm** [comment]
Great list, thanks, going to implement a few things when I have time ☺️

**Comment by rmnblm** [comment]
Today I could only work on some of these offline features and it's still WIP. 
I'm going to work on it as soon as I have time again.

**Comment by nitrag** [comment]
Testing the code you've committed so far, great! Is it a bug that I can re-download the same area over and over? Seems to me, to save bandwidth, the user shouldn't have to re-download the same content twice and should be skipped.

Also downloading a large area, it only downloads 1284 resources before halting. Pausing and Resuming starts the whole thing over. EDIT: Opened https://github.com/mapbox/mapbox-gl-native/issues/6814

Are these mapbox sdk errors?

**Comment by rmnblm** [comment]
Sadly I didn't have much time left to work on this repo the last few days. 

There is already a mechanism to prevent the user to download the same content over and over again, it just isn't clearly when using my example. When you press download, Mapbox "downloads" the content from the disk and not over network. 

> Your Mapbox-powered application will reuse tiles and resources that are required by multiple regions, conserving network traffic and disk space.

This is the same as stated above. Pausing and resuming assumes that it downloads the whole thing over but it doesn't. It first "downloads" the content from disk and then starts downloading the remaining tiles over network. (I've read this somewhere sometime but I couldn't find it anymore)

Well, that's how I understood how the whole process works based on the [documentation](https://www.mapbox.com/help/mobile-offline/). Maybe @1ec5 can confirm this?

**Comment by 1ec5** [comment]
> Pausing and resuming assumes that it downloads the whole thing over but it doesn't. It first "downloads" the content from disk and then starts downloading the remaining tiles over network. (I've read this somewhere sometime but I couldn't find it anymore)

That’s correct. This behavior is documented in the [MGLOfflinePack documentation](https://www.mapbox.com/ios-sdk/api/3.4.0-beta.1/Classes/MGLOfflinePack.html#/c:objc%28cs%29MGLOfflinePack%28im%29resume):

> When a pack resumes after being suspended, it may begin by iterating over the already downloaded resources. As a result, the `progress` structure’s `countOfResourcesCompleted` field may revert to 0 before rapidly returning to the level of progress at the time the pack was suspended.

mapbox/mapbox-gl-native#5808 tracks making this behavior more intuitive.
